include/seqan3/search/kmer_index/detail/shape_iterator.hpp

```diff
@@ -0,0 +1,461 @@
+     * \{
+     */
+    shape_iterator()                                   = default; //!< Defaulted.
+    shape_iterator(shape_iterator const &)             = default; //!< Defaulted.
+    shape_iterator(shape_iterator &&)                  = default; //!< Defaulted.
+    shape_iterator & operator=(shape_iterator const &) = default; //!< Defaulted.
+    shape_iterator & operator=(shape_iterator &&)      = default; //!< Defaulted.
+    ~shape_iterator()                                  = default; //!< Defaulted.
+    
+    /*!\brief Construct from a given iterator on the text and a seqan3::shape.
+    * /param[in] it_start Iterator pointing to the first position of the text.
+    * /param[in] s_ The seqan3::shape that determines which positions participate in hashing.
+    *
+    * ### Complexity
+    *
+    * Linear in size of shape.
+    */
+    shape_iterator(it_t it_start, shape s_) : 
+        shape_{s_}, text_start{it_start}, text_left{it_start}, text_right{it_start}
+    {
+        assert(std::ranges::size(shape_) > 0);
+        ungapped_shape = std::ranges::all_of(shape_, [](bool b) { return b; } );
+
+        roll_factor = std::pow(sigma, std::ranges::size(shape_) - 1);
+
+        hash_full();
+    }
+    //!\}
+
+    //!\anchor shape_iterator_comparison
+    //!\name Comparison operators
+    //!\{
+    //\!brief Compare to iterator on text.
+    friend bool operator==(shape_iterator const & lhs, sentinel_t const & rhs) noexcept
+    {
+        return lhs.text_right == rhs;
+    }
+
+    //\!brief Compare to iterator on text.
+    friend bool operator==(sentinel_t const & lhs, shape_iterator const & rhs) noexcept
+    {
+        return lhs == rhs.text_right;
+    }
+
+    //\!brief Compare to another shape_iterator.
+    friend bool operator==(shape_iterator const & lhs, shape_iterator const & rhs) noexcept
+    {
+        return std::tie(lhs.text_right, lhs.shape_) == std::tie(rhs.text_right, rhs.shape_);
+    }
+
+    //\!brief Compare to iterator on text.
+    friend bool operator!=(shape_iterator const & lhs, sentinel_t const & rhs) noexcept
+    {
+        return !(lhs == rhs);
+    }
+
+    //\!brief Compare to iterator on text.
+    friend bool operator!=(sentinel_t const & lhs, shape_iterator const & rhs) noexcept
+    {
+        return !(lhs == rhs);
+    }
+
+    //\!brief Compare to another shape_iterator.
+    friend bool operator!=(shape_iterator const & lhs, shape_iterator const & rhs) noexcept
+    {
+        return !(lhs == rhs);
+    }
+
+    //\!brief Compare to another shape_iterator.
+    friend bool operator<(shape_iterator const & lhs, shape_iterator const & rhs) noexcept
+    {
+        if (lhs.shape_ != rhs.shape_)
+            return false;
+        return lhs.text_right < rhs.text_right;
+    }
+
+    //\!brief Compare to another shape_iterator.
+    friend bool operator>(shape_iterator const & lhs, shape_iterator const & rhs) noexcept
+    {
+        if (lhs.shape_ != rhs.shape_)
+            return false;
```

Comment:
your implementation of < and > is a little weird I think because usually, you would expect: if !(a < b) then a >= b.. but your operators always return false if the shapes are unequal. I don't know if this kind of comparison is useful if there is the possibility that the shape is different (we cannot check this at compile time)...

Reply:
True, I should probably at least compare the shape sizes.